### PR TITLE
Add Mastercode SubGHz Protocol

### LINKED ISSUE #3182: Add new SubGhz protocol for Clemsa MasterCode MV12
### Description of the feature you're suggesting.

I have a Clemsa Mastercode MV12 garage remote that its not supported now in flipper zero. 
Its a fixed code 433.92 MHz remote with modulation AM270.

Testing with flipperzero i have found that its very similar to Clemsa protocol already implemented, but with some changes:

- Time of the On, Off, pause.
- Length of the key.
- Order of the bits for switches.  

I attach the SUB files for te button 1 and button 2.
The DIP configuration is :
  
(+): ---**---
(o): --------
(-): \*\*\*--\*\*\*

I have already implemented the new protocol and could create a MR

[Mv12_B1.zip](https://github.com/flipperdevices/flipperzero-firmware/files/13226175/Mv12_B1.zip)

617EPLOnN9L _AC_SL1024_

mando-de-garaje-clemsa-mastercode-mv-12

### Anything else?

I have already implemented the new protocol and could create a MR

### PR DESCRIPTION
**Pull request by flipperzelebro**
# What's new
Resolves #3182

Implements the Sub GHz Protocol for the garage remote Clemsa Mastercode MV12 .

Its a fixed code 433.92 MHz remote with modulation AM270.

Its very similar to Clemsa protocol already implemented, but with some changes:

- Time of the On, Off, pause.
- Length of the key.
- Order of the bits for switches.

# Verification 

The PR includes a raw SUB file to  test the decode and a SUB to test de codification.
Attached are 2 additional SUB to test it.
In the wild you can test it using a remote and using 433.92 MHz and modulation AM270.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by Skorpionm** [comment]
In your RAW file the frequency is set to 434.075 MHz, I assume that this is not the correct frequency. The RAW file itself contains no payload, only noise
image

**Comment by Skorpionm** [comment]
You have 36bit in your parcel, you do not display and decrypt everything. It is possible to obtain RAW recordings of different switch positions for analysis. there's clearly enough tone here

**Comment by flipperzelebro** [comment]
It´s true you are righ,!!!  Now I have commited 2 correct raw files and 2 tests for them.

The fisrt file mastercode_raw.sub has this DIP setting:
(+): ---**---
(o): --------
(-): \*\*\*--\*\*\*

and the second mastercore2_raw.sub  has:
(+): --***---
(o): --------
(-): \*\*---\*\*\* 

Also I fixed the key printed in info to show all the hex digits instead of the fisrts 2 bytes.

File: lib/subghz/protocols/mastercode.c [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+                (DURATION_DIFF(
+                     instance->decoder.te_last, subghz_protocol_mastercode_const.te_long) <
+                 subghz_protocol_mastercode_const.te_delta * 8) &&
+                (DURATION_DIFF(duration, subghz_protocol_mastercode_const.te_short) <
+                 subghz_protocol_mastercode_const.te_delta)) {
+                subghz_protocol_blocks_add_bit(&instance->decoder, 1);
+                instance->decoder.parser_step = MastercodeDecoderStepSaveDuration;
+            } else if(
+                //DURATION_DIFF(duration, subghz_protocol_mastercode_const.te_short * 51) <
+                //subghz_protocol_mastercode_const.te_delta * 25) {
```
Comment (Skorpionm) [inline lib/subghz/protocols/mastercode.c]: please delete comment

File: applications/debug/unit_tests/resources/unit_tests/subghz/mastercode2_raw.sub [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Filetype: Flipper SubGhz RAW File
```
Comment (Skorpionm) [inline applications/debug/unit_tests/resources/unit_tests/subghz/mastercode2_raw.sub]: delete the file it doesn't work, too little data recorded

File: applications/debug/unit_tests/subghz/subghz_test.c [hunk 3]
```diff
@@ -654,6 +654,20 @@ MU_TEST(subghz_decoder_kinggates_stylo4k_test) {
...
 }
 
+MU_TEST(subghz_decoder_mastercode_test) {
+    mu_assert(
+        subghz_decoder_test(
+            EXT_PATH("unit_tests/subghz/mastercode_raw.sub"), SUBGHZ_PROTOCOL_MASTERCODE_NAME),
+        "Test decoder " SUBGHZ_PROTOCOL_MASTERCODE_NAME " error\r\n");
+}
+
+MU_TEST(subghz_decoder_mastercode_test2) {
```
Comment (Skorpionm) [inline applications/debug/unit_tests/subghz/subghz_test.c]: a test with 1 file, but working, is enough

**Comment by Skorpionm** [comment]
and again, please give me, just in the archive, recordings of the remote control with different switch positions, the recording should be with the button pressed for about 2+ seconds. switch positions (all down, all in the center, all up, all ladder) and you also need a photo of the DIP switches on the remote control, something tells me that you missed something

File: applications/debug/unit_tests/resources/unit_tests/subghz/mastercode2_raw.sub (hunk 2 above)
Comment (flipperzelebro) [reply applications/debug/unit_tests/resources/unit_tests/subghz/mastercode2_raw.sub]: Deleted in then new commit

File: applications/debug/unit_tests/subghz/subghz_test.c (hunk 3 above)
Comment (flipperzelebro) [reply applications/debug/unit_tests/subghz/subghz_test.c]: Ok, deleted the test2.

Comment (flipperzelebro) [reply applications/debug/unit_tests/subghz/subghz_test.c]: Ok, I removed the test2

**Comment by flipperzelebro** [comment]
No  problem, I really appreciate you check it, because its my first contribution and its fine to double checki it.
I wish the files are fine 
IMG_20231103_182217
[SUB mastercode.zip](https://github.com/flipperdevices/flipperzero-firmware/files/13253043/SUB.mastercode.zip)

**Comment by Skorpionm** [comment]
image

1. you lose the older one when displaying what you need
B 7E 00 3C 08
displayed
0 7E 00 3C 08

2. unit_test doesn't work.
image

you can run and check, assemble and download
"./fbt flash_usb_full FIRMWARE_APP_SET=unit_tests"
launch CLI cmd "unit_tests subghz"

3. Bit durations and generation of guard time during emulation were incorrectly calculated
image
    .te_short = 1072,
    .te_long = 2145,
    
    P.S. don't get me wrong, I'm not being picky, we need a fully working PR

**Comment by flipperzelebro** [comment]
Finally I have fixed the two bugs that I had, on the one hand the duration of the pulses was not correct and on the other hand the Key was not being displayed completely.    
I found @Skorpionm  comments and the detail on how to do the unit tests very helpful. I think everything is now correct.

**Review by Skorpionm — APPROVED** [review]

**Comment by Skorpionm** [comment]
@flipperzelebro  
SUCCESS! I'm very glad you didn't give up.

**Review by skotopes — APPROVED** [review]

**Pull request merged**